src/SmartComponents/modules/reducers.js

```diff
@@ -6,9 +6,66 @@ const initialState = {
     selectedSystemIds: [],
     filterDropdownOpened: false,
     stateFilter: 'all',
-    factFilter: ''
+    factFilter: '',
+    totalFacts: undefined,
+    page: 1,
+    perPage: 10,
+    filteredData: {}
 };
 
+function getTotal(data, stateFilter, factFilter) {
+    let total = 0;
+
+    for (let i = 0; i < data.facts.length; i++) {
+        if (data.facts[i].name.includes(factFilter)) {
+            if (stateFilter.toLowerCase() === 'all' || stateFilter === undefined) {
+                total += 1;
+            }
+            else if (stateFilter === data.facts[i].state) {
+                total += 1;
+            }
+        }
+    }
+
+    return total;
+}
+
+function filterCompareData(data, selectedPage, factsPerPage, stateFilter, factFilter) {
+    if (data === undefined || data.facts === undefined) {
+        return {};
+    }
+
+    let total = 0;
+    let filteredData = {};
+    let facts = [];
+
+    for (let i = 0; i < data.facts.length; i += 1) {
+        if (data.facts[i].name.includes(factFilter)) {
+            if (stateFilter.toLowerCase() === 'all' || stateFilter === undefined) {
+                total += 1;
+                if (Math.ceil(total / factsPerPage) === selectedPage) {
+                    facts.push(data.facts[i]);
+                }
+            }
+            else if (stateFilter === data.facts[i].state) {
+                total += 1;
+                if (Math.ceil(total / factsPerPage) === selectedPage) {
+                    facts.push(data.facts[i]);
+                }
+            }
+        }
+    }
```

Comment:
I think this could be cleaner to split into two distinct stages with their own functions:

 1) Filter all the data
 2) Paginate the data

Reply:
This will allow you to unit test cleaner as well. As you can varify pagination logic and filtering logic keeping them more robust over time.